### ISSUE TITLE
Getting a lot of "Broadcasts disabled" error logs when not using broadcasts

### ISSUE THREAD

**Issue by jeromegn** [caio/foca#18]
I've turned off broadcasts through `foca` entirely a while ago and do my own thing now.

I'm not using `Foca::broadcast` or `Foca::add_broadcast` anywhere in my code, yet I keep seeing "Broadcasts disabled" error logs.

Can it happen any other way? Perhaps this is due to my corrupted messages being interpreted as custom broadcasts by foca?

**Comment by jeromegn** [comment]
Looking deeper, this could definitely be due to corrupt messages.

If any buffer contains more bytes after parsing SWIM-related messages, they're considered a broadcast and `foca` will try to `handle_broadcast`.

I've been fixing a few issues where the byte payloads I sent weren't quite right and it seemed to help with various `foca` errors, but I still see a lot of "Broadcasts disabled" errors.

I was trying to think if it could be due to UDP packets getting fractioned? I feel like that would cause a different error.

These errors seem to happen at very specific intervals, which is odd to me. Looks like "twice within 200ms every second".

<img width="208" alt="image" src="https://user-images.githubusercontent.com/43325/205916733-3ff3a9a6-48a9-4a40-80bb-8459fe231e13.png">

**Comment by caio** [comment]
What a weird thing!

Are you seeing this on every node or just the one?

> I was trying to think if it could be due to UDP packets getting fractioned? I feel like that would cause a different error.

That wouldn't be the cause indeed as the fragmentation would be handled in the ip layer i reckon- The main reason to stick to packets smaller than the MTU is to minimize loss (if your packet loss is 1%, a fragmented message has 2% of failing)

These "Broadcasts disabled" errors you are seeing are indicative of foca receiving more data than it should during `handle_data`. Because of the design of appending data to the tail of the packet, foca simply assumes that if there's trailing data if must be a broadcast.

> These errors seem to happen at very specific intervals, which is odd to me. Looks like "twice within 200ms every minute at around the 30s mark".

Interesting. Sounds like something is happening periodically and receiving 2 replies. `Config::new_wan` sets periodic announce to 60s with a fan-out of 2, perhaps not a coincidence? :thinking: 

The first thing that comes to mind is an outdated client: it's sending messages periodically and receiving a reply that it can't handle (eg: it's sending a periodic announce and receiving a notification that they're down; but the node is on a previous foca version that doesn't know about this new notification message). Could probably explain why we keep seeing the weird off-by-one counts on the other issue.

If that were the case I think we'd be seeing occasional decode error messages in the logs. Are there more errors in the log that look suspicious?

---

I don't have any other theory besides this one :sweat_smile: so if that's not it, maybe we can try to get some more data from these failures: write a copy of the [NoCustomBroadcast handler](https://docs.rs/foca/latest/src/foca/lib.rs.html#1425) that emits more info such as:

- what's the length of this trailing data
- if we're misinterpreting packets due to different encoding versions maybe the trailing data is a bunch of cluster updates? try using your codec's `decode_member` to see if it works
- maybe there's a very old client still sending custom broadcasts? try to decode it using your old broadcast handler `receive_item`

foca's error reporting / logging is unfortunately pretty shallow right now, especially when not running on debug level, so figuring out who is the bad actor (if any) may entail adding logs before giving the payload to foca- at least until I get to improving this sorry state

**Comment by jeromegn** [comment]
I've added a simple debug broadcast handler and I've found that all the broadcasts it gets **have a length of zero**.

Implementation:

```rust
#[derive(Debug, Clone, Copy)]
pub struct DebugCustomBroadcasts;

impl<T> BroadcastHandler<T> for DebugCustomBroadcasts {
    type Broadcast = &'static [u8];
    type Error = BroadcastsDisabledError;

    fn receive_item(&mut self, mut data: impl Buf) -> Result<Option<Self::Broadcast>, Self::Error> {
        let mut buf = vec![];
        data.copy_to_slice(&mut buf);

        warn!("received custom broadcast of len: {}", buf.len());
        println!("to string: {}", String::from_utf8_lossy(&buf));
        println!("to vec: {buf:?}");

        Err(BroadcastsDisabledError)
    }
}
```

What I see in my logs:

```
received custom broadcast of len: 0
to string:
to vec: []
```

**Comment by jeromegn** [comment]
(Disregarded my now-deleted comment, which you probably got an email about)

I added a "debug" handler and noticed the following about the extra bytes format, but I can't quite figure out where they're coming from:
- `16u8` representing the length of the field value
- a uuid (that's what I use for my node ids)
- `19u8` or `20u8`, representing the length of a string
- the hostname
- bytes I don't understand: `[1, 252, 1]`
- an IPv4 address (sometimes, but rarely, cut off) - the one set for gossip
- then usually a bunch of `[0,0,0, ...]`, but sometimes a bit more, incomplete, data.

It _looks_ a lot like my `Identity` bincode encoded?

```rust
#[derive(Clone, PartialEq, Eq, Deserialize, Serialize)]
pub struct Actor {
    id: ActorId,
    name: ActorName,
    addr: SocketAddr,
    region: String,
    datacenter: String,
    write_mode: WriteMode,
    instance_id: Uuid,

    // An extra field to allow fast rejoin
    bump: u16,
}
```

Where:
- `ActorId` is a wrapper for `Uuid`
- `ActorName` is a wrapper for a `CompactString`
- `region` is usually 3 characters, and I can sometimes see it in the extra data at the end, but it's usually cut off
- `datacenter` is often `region + number`, and I do sometimes see both at the end of the data (if that's present), like: `[3, 111, 114, 100, 4, 111, 114, 100] => "ord" "ord1"`

However, I've never seen `write_mode`, `instance_id` or `bump`.

Other interesting stuff:
- I don't send my Identity (`Actor`) with any of my custom payloads.
- I'm prefacing my payloads with `0u8` when it's a `foca` payload (the only type of message that sends the `Actor`).

Function that sends `foca` messages:

```rust
async fn handle_gossip_to_send(socket: Arc<UdpSocket>, mut to_send_rx: UnboundedReceiver<(Actor, Bytes)>) {
    let mut buf = BytesMut::with_capacity(FRAGMENTS_AT);

    while let Some((actor, payload)) = to_send_rx.recv().await {
        let addr = actor.addr();
        let name = actor.name().clone();
        buf.put_u8(0);
        buf.put_slice(&payload);

        let bytes = buf.split().freeze();
        let socket = socket.clone();
        spawn_counted(async move {
            match socket.send_to(bytes.as_ref(), addr).await {
                Ok(n) => {
                    histogram!("corrosion.gossip.sent.bytes", n as f64, "actor" => name.0.to_string());
                }
                Err(e) => {
                    error!("could not send SWIM message via udp to {addr}: {e}");
                }
            }
        });
        increment_counter!("corrosion.gossip.send.count", "actor" => actor.name().0.to_string());
    }
}
```

And this is how we parse it:

```rust
// receiving on the socket

async move {
            let mut recv_buf = vec![0u8; FRAGMENTS_AT];

            let actor_id = actor.id();
            loop {
                match socket.recv_from(&mut recv_buf).await {
                    Ok((len, from_addr)) => {
                        trace!("udp received len: {len}");
                        let recv = recv_buf[..len].to_vec().into();
                        let foca_tx = foca_tx.clone();
                        let disseminated = disseminated.clone();
                        let gossip_msg_tx = gossip_msg_tx.clone();
                        let bcast_priority_tx = bcast_priority_tx.clone();
                        tokio::spawn(async move {
                            let mut codec = LengthDelimitedCodec::builder().length_field_type::<u32>().new_codec();
                            let mut databuf = BytesMut::new();
                            if let Err(e) = handle_payload(
                                None,
                                &mut databuf,
                                &mut codec,
                                recv,
                                &foca_tx,
                                actor_id,
                                &disseminated,
                                &gossip_msg_tx,
                                &bcast_priority_tx,
                            ) {
                                error!("could not handle UDP payload from '{from_addr}': {e}");
                            }
                        });
                    }
                    Err(e) => {
                        error!("error receiving on gossip udp socket: {e}");
                        sleep(Duration::from_secs(1)).await;
                    }
                }
            }
        }

// handling the data

pub fn handle_payload(
    mut kind: Option<PayloadKind>,
    buf: &mut BytesMut,
    codec: &mut LengthDelimitedCodec,
    mut payload: Bytes,
    foca_tx: &UnboundedSender<FocaInput>,
    actor_id: ActorId,
    disseminated: &Disseminated,
    gossip_msg_tx: &GossipMessageTx,
    bcast_priority_tx: &GossipMessageTx,
) -> eyre::Result<PayloadKind> {
    let kind = kind.take().unwrap_or_else(|| {
        let kind = match payload.get_u8() {
            0 => PayloadKind::Swim,
            1 => PayloadKind::Broadcast,
            2 => PayloadKind::PriorityBroadcast,
            n => PayloadKind::Unknown(n),
        };
        increment_counter!("corrosion.payload.recv.count", "kind" => kind.to_string());
        kind
    });

    debug!("got payload kind: {kind:?}");

    match kind {
        // SWIM
        PayloadKind::Swim => {
            trace!("received SWIM gossip");
            // And simply forward it to foca
            _ = foca_tx.send(FocaInput::Data(payload));
        }
        // ...
    }

    Ok(kind)
}

```

The codec is only used if we have a broadcast, they're length-delimited.

Any ideas?

**Comment by caio** [comment]
Thanks a lot for all the info!

I agree, looks like these extra bytes contain a representation of your identity. I think it's enough evidence to assume for now that these are cluster updates (the extra bytes you couldn't interpret may be encoded enum variants, foca::State, foca::Incarnation)

The code you shared looks good to me, so I think we need to start verifying assumptions to reduce the problem space. Here's what I'd do:

1. Create a new PayloadKind::SwimV2: this would isolate members running new code from ones running old code; Would rule out my main worry: a member with an older foca version or a different bincode config still talking to the cluster
2. Along with the payloadkind tag, I'd put a u16 with the actual length of the payload: would allow us to verify the assumption about fragmentation not being a problem (you could check the length from `recv` vs this new one to ensure that the length is correct, for example)
3. And maybe add a checksum (or crypto signature) to it too. I find it unlikely that a bit-flipping-cosmic-ray-kind-of-thing would cause corruption consistently and often enough to lead to what you're seeing; but maybe not? maybe there's a chance of actual adversarial input?

**Comment by jeromegn** [comment]
Good stuff! I'll do that when I find some time to work on it more again.

**Comment by jeromegn** [comment]
I haven't had time to implement the whole troubleshooting shebang here, but I did try something: I started logging from which node the data causing these error originated (not necessarily the one that produced the data, but the one that sent it).

It always happens over UDP, so probably only when sending "SWIM data" or small custom broadcasts (outside of the foca broadcasting functionality, of course).

I found that it's pretty random.

On thing I did notice is that the time at which these happen changed when I restarted the cluster. The node I was watching logs 2 "broadcasts disabled" errors per minute, but the second at which it logs them suddenly changed as it was restarted.

For example:

```
$ systemctl status corrosion
● corrosion.service - corrosion
   Loaded: loaded (/etc/systemd/system/corrosion.service; enabled; vendor preset: enabled)
   Active: active (running) since Fri 2022-12-30 21:35:26 UTC; 7min ago

...

Dec 30 21:33:58 dev-ovh-hil1-a09b corrosion[703421]: {"timestamp":"2022-12-30T21:33:58.791746Z","level":"ERROR","fields":{"message":"error handling foca data from Udp([fc01:a7b:5f73::]:7878): Broadcasts disabled"},"target":"corrosion::gossip"}
Dec 30 21:33:58 dev-ovh-hil1-a09b corrosion[703421]: {"timestamp":"2022-12-30T21:33:58.948470Z","level":"ERROR","fields":{"message":"error handling foca data from Udp([fc01:a7b:d6e9::]:7878): Broadcasts disabled"},"target":"corrosion::gossip"}
Dec 30 21:34:58 dev-ovh-hil1-a09b corrosion[703421]: {"timestamp":"2022-12-30T21:34:58.734984Z","level":"ERROR","fields":{"message":"error handling foca data from Udp([fc01:a7b:307::]:7878): Broadcasts disabled"},"target":"corrosion::gossip"}
Dec 30 21:34:58 dev-ovh-hil1-a09b corrosion[703421]: {"timestamp":"2022-12-30T21:34:58.789960Z","level":"ERROR","fields":{"message":"error handling foca data from Udp([fc01:a7b:17ec::]:7878): Broadcasts disabled"},"target":"corrosion::gossip"}
Dec 30 21:35:26 dev-ovh-hil1-a09b corrosion[735300]: {"timestamp":"2022-12-30T21:35:26.523364Z","level":"ERROR","fields":{"message":"error handling foca data from Udp([fc01:a7b:d203::]:7878): Broadcasts disabled"},"target":"corrosion::gossip"}
Dec 30 21:35:26 dev-ovh-hil1-a09b corrosion[735300]: {"timestamp":"2022-12-30T21:35:26.526571Z","level":"ERROR","fields":{"message":"error handling foca data from Udp([fc01:a7b:b60f::]:7878): Broadcasts disabled"},"target":"corrosion::gossip"}
Dec 30 21:35:26 dev-ovh-hil1-a09b corrosion[735300]: {"timestamp":"2022-12-30T21:35:26.601093Z","level":"ERROR","fields":{"message":"error handling foca data from Udp([fc01:a7b:17ec::]:7878): Broadcasts disabled"},"target":"corrosion::gossip"}
Dec 30 21:36:26 dev-ovh-hil1-a09b corrosion[735300]: {"timestamp":"2022-12-30T21:36:26.586608Z","level":"ERROR","fields":{"message":"error handling foca data from Udp([fc01:a7b:9d35::]:7878): Broadcasts disabled"},"target":"corrosion::gossip"}
Dec 30 21:36:26 dev-ovh-hil1-a09b corrosion[735300]: {"timestamp":"2022-12-30T21:36:26.708542Z","level":"ERROR","fields":{"message":"error handling foca data from Udp([fc01:a7b:6a::]:7878): Broadcasts disabled"},"target":"corrosion::gossip"}
```

So it used to get them at the `58`th second and now it gets it at the `26`th second, which matches the second at which it started last time.

I don't know what the heck :D I'll try more involved troubleshooting after the holidays.

**Comment by caio** [comment]
Aha, I found what's up! We were on the right track when I wrote this:

> Interesting. Sounds like something is happening periodically and receiving 2 replies. Config::new_wan sets periodic announce to 60s with a fan-out of 2, perhaps not a coincidence? 🤔 

I forgot to dig into it, but now with these logs you shared my mind snapped back to the right train of thought:

With this correlation with the start time I became fairly certain that it's related to the Announce/Feed cycle; Then I went to the code and found a [note that I should've addressed before releasing the feature](https://github.com/caio/foca/blob/c7145ae70cd1866f2bb4c2500db98c3d63ad373b/src/lib.rs#L1238)

I think bincode length-prefixes everything that's not fixed length, so this shouldn't be the cause for #19 ... but maybe?

The fix with a reproducing test is here: aa06a52c2d1111a934d35dc0b04aee736ad15757 it's really trivial- I'll ship it soon

The main reason I didn't detect this is because my cluster never sends large enough messages (tiny ids, very few members, default Config::max_packet_size) and my codec is super careful (easy to achieve with a small identity and not using varints)

Aside: As I fixed this I noticed that foca was allocating a lot more than it should- it's because I confused `BufMuf::split` with `BufMut::split_off`, so it's never reusing the buffers. I'll fix this one in a separate release just in case actually reusing buffers surfaces more bugs 😬 

Thanks a lot for all the info and sorry for all the head scratching you might've done debugging on your side haha

I'll close this issue when I land the release, but feel free to reopen in case the error doesn't disappear.

**Comment by jeromegn** [comment]
Deploying! That does sound like the fix though :)

**Comment by jeromegn** [comment]
That fixed it :)

It did not fix #19 unfortunately, would've been nice haha.

**Comment by caio** [comment]
hahaha agreed!
at least it's one less bug for 2023 🎉
